### ISSUE TITLE
Wrong source code stepping info

### ISSUE THREAD

**Issue by anass-b** [dotnet/roslyn#16747]
**Original Issue**: https://github.com/dotnet/coreclr/issues/8958

**Version Used**: VS 2017 RC, VS 2015 Update 3, and dotnet CLI 1.0.0-preview2-1-003177

**Steps to Reproduce**:

Given the following code:
```c#
using System;
using System.Threading.Tasks;

class Program
{
    static void Main() { TestAsync().Wait(); }

    static async Task TestAsync()
    {
        int nonZero = await FalseAsync();

        // The problem is that the debugger steps into the body of the if below but it should not.
        if (nonZero == 0)
        {
            throw new Exception();
        }
    }

    static async Task<int> FalseAsync() { return await Task.FromResult(42); }
}
```
1. From Visual Studio or VS Code, put a breakpoint on this line `if (nonZero == 0)`, check the `nonZero` variable value, you will see it's 42
2.  Step over with the debugger

**Expected Behavior**:
The debugger steps out of the `if`

**Actual Behavior**:
The debugger steps inside the `if`

**Screenshot**:
![47389342-db96-11e6-94b8-3f16c869d7ca](https://cloud.githubusercontent.com/assets/8070594/22303888/8a258268-e335-11e6-96a5-bda5cf96a2b0.png)

**Screen Capture of Disassembly + Registers**
[debug.zip](https://github.com/dotnet/roslyn/files/730668/debug.zip)

**Comment by tmat** [comment]
@jcouv Looks similar to what we were looking into recently.

**Comment by jcouv** [comment]
Yes, this is probably the same issue. https://github.com/dotnet/roslyn/issues/12564
The code that is executed is correct, but the debugger provides misleading highlights, which makes it look like the branch was entered (when in fact it was not).

**Comment by gafter** [comment]
This is a dup of #12915. #12564 may be related too.

**Comment by jcouv** [comment]
Fixed by #16803 (Missing sequence point on implicit return in async method)
The fix unfortunately didn't make it into RTW, but will be released in the next ship window.
